runtime/array.c

```diff
@@ -589,3 +589,41 @@ CAMLprim value caml_array_concat(value al)
   }
   return res;
 }
+
+CAMLprim value caml_array_fill (value array, value v_ofs, value v_len, value val)
+{
+  value old;
+  intnat ofs = Long_val(v_ofs);
+  intnat len = Long_val(v_len);
+  value* fp;
+
+  /* See caml_modify for invariants */
+
+#ifdef FLAT_FLOAT_ARRAY
+  if (Tag_val(array) == Double_array_tag) {
+    double d = Double_val (val);
+    for (; len > 0; len--, ofs++)
+      Store_double_flat_field(array, ofs, d);
+    return Val_unit;
+  }
+#endif
+  fp = &Field(array, ofs);
+  if (Is_young(array)) {
+    for (; len > 0; len--, fp++) *fp = val;
+  } else {
+    CAMLassert(Is_in_heap(fp));
+    int is_val_young_block = Is_block(val) && Is_young(val);
+    for (; len > 0; len--, fp++) {
+      old = *fp;
+      if (old == val) continue;
+      *fp = val;
+      if (Is_block(old)) {
+        if (Is_young(old)) continue;
+        if (caml_gc_phase == Phase_mark) caml_darken(old, NULL);
+      }
+      if (is_val_young_block) add_to_ref_table (&caml_ref_table, fp);
+    }
+    /* Should we call caml_check_urgent_gc? */
```

Comment:
This is what `caml_make_vect`, `caml_make_array`,  `caml_array_blit`, `caml_array_gather` the last two with the comment:
````
  /* Many {caml_modify, caml_initialize} in a row can create a lot of old-to-young refs.
     Give the minor GC a chance to run if it needs to. */
 ````

Reply:
Ah, I see! Makes sense, though I expect @damiendoligez should have the final answer :)

Reply:
@damiendoligez, or another GC expert: do you have an opinion on where calling caml_check_urgent_gc is a good idea there?

Reply:
Yes, you should call `caml_check_urgent_gc` here.

Reply:
Thanks.

Can the call be conditioned to `is_val_young_block` (i.e. `Is_block(val) && Is_young(val)`)?

For future reference, can you explain a bit why this is needed here?  The functions does nothing (except for calling `caml_memprof_handle_postponed`) if neither `caml_requested_major_slice` nor `caml_requested_minor_gc` are set.  But AFAICT, those flags cannot be set during the execution of `caml_array_fill` (the only function calls are `caml_darken` and `add_to_ref_table`, but I could not find a code path from them leading to setting one of the flags).  So this would mean the flags were already set before?

Reply:
Suggestion merged, but still interested in the explanation @damiendoligez

Reply:
Unless I'm mistaken, `add_to_ref_table` can call `realloc_ref_table`, which can call `caml_request_minor_gc`, which sets `caml_requested_minor_gc`.

You don't want to use arbitrary amounts of memory for the remembered set (ref_table) so when it gets too large you trigger a minor GC to eliminate it.

Reply:
Ah, sorry, I don't know how I missed that.  (Side note: shouldn't `caml_request_minor_gc` and `caml_requested_minor_gc`be moved from signals.c to e.g. minor_c?  They are not really signal-related.)

Reply:
In fact, you could do `if (is_val_young_block) caml_check_urgent_gc (Val_unit);`

Reply:
Ok, done.